pkg/zson/sort.go

```diff
@@ -12,37 +12,66 @@ type SortFn func(a *Record, b *Record) int
 // Internal function that compares two values of compatible types.
 type comparefn func(a, b []byte) int
 
-func rawcompare(a, b []byte, dir int) int {
-	v := bytes.Compare(a, b)
-	if v < 0 {
-		return -dir
-	} else {
-		return dir
+func isUnset(typ zeek.Type, val []byte) bool {
+	if val == nil || zeek.SameType(typ, zeek.TypeNone) || zeek.SameType(typ, zeek.TypeUnset) {
+		return true
 	}
+	return false
 }
 
-func NewSortFn(dir int, fields ...string) SortFn {
+// NewSortFn() creates a function that compares a pair of Records
+// based on the provided ordered list of fields.
+// The returned function uses the same return conventions as standard
+// routines such as bytes.Compare() and strings.Compare(), so it may
+// be used with packages such as heap and sort.
+// A record in which one of the comparison fields is not present is
+// considered to be smaller than a record in which the field is present.
+// The handling of records in which a comparison field is unset is
+// governed by the unsetMax parameter.  If this parameter is true,
+// a record with unset is considered larger than a record with any other
+// value, and vice versa.
+func NewSortFn(unsetMax bool, fields ...string) SortFn {
 	sorters := make(map[*zeek.Type]comparefn)
 	return func(a *Record, b *Record) int {
 		for _, field := range fields {
 			vala, typea, erra := a.Access(field)
 			valb, typeb, errb := b.Access(field)
 
 			// Errors indicate the field isn't present, sort
-			// these records last
+			// these records to the minimum value so they
+			// appear first in sort output.
 			if erra != nil && errb != nil { return 0 }
-			if erra != nil { return 1 }
-			if errb != nil { return -1 }
+			if erra != nil { return -1 }
+			if errb != nil { return 1 }
+
+			// Handle unset according to unsetMax
+			unsetA := isUnset(typea, vala)
+			unsetB := isUnset(typeb, valb)
+			if unsetA && unsetB { return 0 }
+			if unsetA {
+				if unsetMax {
```

Comment:
I'm not sure if this is cosmetic suggestion is idiomatic but maybe do return m[unsetMax] where m is a map[bool]int ?

Reply:
(and -1 * m[unsetMax] below)

Reply:
Whoops, meant to reply to this before closing.  I find this approach less readable but I can switch to it if others really like it.